packages/workbox-webpack-plugin/src/lib/get-webpack-manifest-entries.js

```diff
@@ -0,0 +1,76 @@
+const url = require('url');
+
+const entry = (revision, filePath, publicPath) => ({
+  revision,
+  url: url.resolve(publicPath, filePath),
+});
+
+/**
+ * Filter chunks that have one of the provided names (from config.chunks)
+ * @param {Array<Object>} chunks Webpack chunks
+ * @param {Array<string>} include Chunk names to include
+ * @param {Array<string>} exclude Chunk names to exclude
+ * @return {Array<Object>} Filtered array of chunks
+ */
+const filterChunks = (chunks, include, exclude) =>
+  chunks.filter((chunk) =>
+    // chunk has a name
+    Object.prototype.hasOwnProperty.call(chunk, 'name')
+      // did not specify include or chunk in include
+      && (!Array.isArray(include) || !!~include.indexOf(chunk.name))
+      // did not specify exclude or chunk not in exclude
+      && (!Array.isArray(exclude) || !~exclude.indexOf(chunk.name))
+  );
+
+/**
+ * Maps webpack asset filenames to their chunk hash
+ * @param {Array<Object>} chunks Webpack chunks
+ * @return {Object} {filename: hash} map
+ */
+const mapAssetsToChunkHash = (chunks) =>
+  chunks.reduce((assetMap, chunk) =>
+      Object.assign(
+        assetMap,
+        ...chunk.files.map((f) => ({[f]: chunk.renderedHash}))
+    ), {});
+
+const generateManifestEntries = (compiler, compilation, config) => {
+  const {hash} = compilation; // compilation hash
+  const {publicPath} = compilation.options.output;
+
+  // Array<string> | undefined
+  const includeNames = config.chunks;
+  const excludeNames = config.excludeChunks;
+
+  // Array<{files: Array<(filename: string)>, name?: string}>
+  let {chunks} = compilation;
+
+  if (includeNames || excludeNames) {
+    // Only include chunks in config.chunks and
+    // exclude any chunks in config.excludeChunks
+    chunks = filterChunks(chunks, includeNames, excludeNames);
+  }
+
+  // Use chunkhash to save a chunk hash to each filename in chunks
+  const assetsToChunkHash = mapAssetsToChunkHash(chunks);
+
+  /**
+   * Files that will be used to generate the manifest entries
+   *
+   * If config.chunks is specifed, we only use files that belong to named chunks
+   * otherwise we use all of webpack's compilation.assets
+   *
+   * @type {Array<string>}
+   */
+  const manifestFiles = includeNames
```

Comment:
I'm kinda confused with all the alterations and this object keys here, would this change to:

```javascript
const assetsToChunkHash = mapAssetsToChunkHash(precacheChunks);
const manifestFiles = Object.keys(precacheChunks);
```

This being due to the check includeNames || excludeNames above rendering this if statement no longer needed.

Reply:
yeah it's kind of confusing. I'll update comments to clarify more.

`assetsToChunkHash` is a map of `filename`: `chunkhash` and only contains assets that belong to a chunk with a name property. It's also filtered out chunks according to `include` / `exclude`.

`compilation.assets` is all assets that belong to the webpack compilation, even assets that belong to chunks with no name.